### PR TITLE
Broadcast auth-loginRequired only once

### PR DESCRIPTION
**Pull request by olivierdeckers**
I have encountered a bug in my angular application which uses your package. It is related to issue #95.
It occurs when using oauth refresh tokens and when multiple calls that all return HTTP 401 are fired in parallel.
The loginRequired event is then fired multiple times, which resulted in my case in multiple refresh_token calls to the oauth endpoint. All but the first call fail as a refresh_token can only be used once. When refreshing the token fails, the app falls back to showing a login form, even though the refresh token was valid and the first call obtained a valid access token.

I could have handled this case in my app by checking that I only fire one refresh_token call in the loginRequired event listener, but since I don't see a use case for emitting the loginRequired event again before loginConfirmed function was called, I thought it would be better to solve it in this library.

In issue #95 you suggest to use one probe call to avoid this issue. This solution is not adequate for my app, because waiting for this probe call would slow the app down unnecessarily in the case where the access token is still valid. This PR eliminates the need to use a probe call by only firing the loginRequired event once.

The event is now only fired if the httpBuffer is empty. If the buffer isn't empty, the the loginRequired event has already been fired, and we can safely assume that the app is in the process of logging in and will call the loginConfirmed method, emptying the buffer.

For me, the current behaviour was unexpected and maybe this change can save others from similar hard-to-reproduce bugs.

### CODE REVIEW

**Comment by witoldsz** [comment]
Hi,
can you rewrite the commit without code reformat, so we could see the actual change in code of this pull request?

File: src/http-auth-interceptor.js [hunk 1]
```diff
@@ -125,6 +129,13 @@
           retryHttpRequest(updater(buffer[i].config), buffer[i].deferred);
         }
         buffer = [];
+      },
+
+      /**
+       * Check whether the buffer is empty
+       */
+      bufferEmpty: function() {
```
Comment (witoldsz) [inline src/http-auth-interceptor.js]: Maybe rename the `bufferEmpty` to `isEmpty` and you could also get rid of that useless comment. What else the `httpBuffer.isEmpty()` can possibly do?

**Comment by olivierdeckers** [comment]
I rolled back the code reformat. Could you take another look?

File: src/http-auth-interceptor.js [hunk 2]
```diff
@@ -56,8 +56,10 @@
             switch (rejection.status) {
               case 401:
                 var deferred = $q.defer();
+                if (httpBuffer.isEmpty()) {
+                  $timeout($rootScope.$broadcast('event:auth-loginRequired', rejection));
```
Comment (witoldsz) [inline src/http-auth-interceptor.js]: Why `$timeout` here?

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "angular-http-auth",
-  "version": "1.2.3",
+  "version": "1.3.0",
```
Comment (witoldsz) [inline package.json]: There will be no version 1.3.0 until release, something more like 1.3.0-SNAPSHOT I guess would do. At least that's what I used to in Java/Maven repositories. I am not sure how does *npm* handle this, though...

Comment (simison) [inline package.json]: It's not a great practise to change this at feature PR's anyway. Version changes and such should be done by the maintainer at separate commits. Cheers!

**Comment by simison** [comment]
@olivierdeckers BTW, you might want to [squash](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) after the work at this PR is done! :-)

File: package.json (hunk 3 above)
Comment (witoldsz) [inline package.json]: Hi @simison 
How do you think it should be done the `npm` way? Should I keep `1.2.3` version at `master` until new release and then upgrade to `1.3.0` or the Maven way to set a `-SNAPSHOT` between releases? Or something else?

Comment (simison) [inline package.json]: Frankly I'm not very familiar with `-SNAPSHOT` stuff.

The way how I've seen it being done (and how I do it), is that once you decide to publish a feature, you push version bumps+changelog to the master and tag a new version at Github to that commit, so that the head of new version is always basically just a version number change. There are also gulp/bash/rake/grunt scripts to do this for you.

You wouldn't want version number changes to package.json within PRs because there might be multiple PR's going on at the same time and once one of them gets merged, you might want to publish new version before merging the next one etc... it gets into a mess to manage quickly. :-)

Similarly  you don't want to have code-style changes, new features, multiple fixes etc within one code. It gets harder for others what actually has been changed. _One thing per PR._

See also [Semver](http://semver.org) to make sure you do proper versioning regarding breaking changes (don't be afraid to bump up the first digit, basically).

Hope this helps! :-)

Comment (witoldsz) [inline package.json]: The thing that seem strange to me is that the version `1.2.3` was published/tagged already. Now we have this pull request. As you say, it should not touch the version - OK, but now we have a new code, new feature, new behavior, but the version `1.2.3` stays untouched and it does not match the released version any more. 

In Java/Maven we keep `x.x.x-SNAPSHOT` during development. Then comes the release, the version gets cleared from snapshot suffix. Then the version gets bumped to the new number, the snapshot gets attached again and we have clear situation on what is the version of each commit.

If I understand correctly, the NPM way is to keep the old version number up until next release, is that correct?

Comment (simison) [inline package.json]: >If I understand correctly, the NPM way is to keep the old version number up until next release, is that correct?

Correct. If people want the latest and greatest, they would just set their package manager to get "#master" (NPM), "@dev" (Composer) or whatever is the practise in their manager. They can also refer to a certain commit hash to get exactly the "snapshot" they want.

So in github version numbers are for pre-releases or releases only.

https://docs.npmjs.com/cli/install

**Comment by witoldsz** [comment]
It was fixed in 1.3.0.
